File: frontend/src/stores/watchlist.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { useApi } from '../composables/useApi'

export const useWatchlistStore = defineStore('watchlist', () => {
  const api = useApi()

  // State
  const symbols = ref([])
  const maxSize = ref(50)
  const loading = ref(false)
  const error = ref(null)

  // Computed
  const count = computed(() => symbols.value.length)
  const isFull = computed(() => count.value >= maxSize.value)

  // Actions
  async function fetchWatchlist() {
    loading.value = true
    error.value = null
    try {
      const data = await api.getWatchlist()
      symbols.value = data.symbols
      maxSize.value = data.max_size
    } catch (e) {
      error.value = e.response?.data?.detail || 'Failed to load watchlist'
      console.error('Failed to fetch watchlist:', e)
    } finally {
      loading.value = false
    }
  }

  async function addSymbol(symbol) {
    loading.value = true
    error.value = null
    try {
      const data = await api.addToWatchlist(symbol)
      symbols.value = data.symbols
      return { success: true, message: data.message }
    } catch (e) {
      const message = e.response?.data?.detail || 'Failed to add symbol'
      error.value = message
      return { success: false, message }
    } finally {
      loading.value = false
    }
  }

  async function removeSymbol(symbol) {
    loading.value = true
    error.value = null
    try {
      const data = await api.removeFromWatchlist(symbol)
      symbols.value = data.symbols
      return { success: true, message: data.message }
    } catch (e) {
      const message = e.response?.data?.detail || 'Failed to remove symbol'
      error.value = message
      return { success: false, message }
    } finally {
      loading.value = false
    }
  }

  async function searchSymbols(query) {
    if (!query || query.length < 1) return []
    try {
      return await api.searchSymbols(query)
    } catch (e) {
      console.error('Failed to search symbols:', e)
      return []
    }
  }

  function isInWatchlist(symbol) {
    return symbols.value.includes(symbol.toUpperCase())
  }

  return {
    // State
    symbols,
    maxSize,
    loading,
    error,
    // Computed
    count,
    isFull,
    // Actions
    fetchWatchlist,
    addSymbol,
    removeSymbol,
    searchSymbols,
    isInWatchlist,
  }
})

File: backend/app/api/routes/watchlist.py
"""
API routes for managing the stock watchlist.
"""
from fastapi import APIRouter, HTTPException, Query, Path
from pydantic import BaseModel

from ...services.watchlist_service import watchlist_service

MAX_SYMBOL_LENGTH = 10

router = APIRouter(prefix="/api/watchlist", tags=["watchlist"])


class WatchlistResponse(BaseModel):
    symbols: list[str]
    count: int
    max_size: int


class SymbolActionResponse(BaseModel):
    success: bool
    message: str
    symbols: list[str]


class SearchResult(BaseModel):
    symbol: str
    name: str
    price: float | None
    in_watchlist: bool


@router.get("", response_model=WatchlistResponse)
async def get_watchlist():
    """Get the current watchlist."""
    symbols = watchlist_service.get_watchlist()
    return WatchlistResponse(
        symbols=symbols,
        count=len(symbols),
        max_size=50
    )


@router.post("/{symbol}", response_model=SymbolActionResponse)
async def add_to_watchlist(
    symbol: str = Path(..., min_length=1, max_length=MAX_SYMBOL_LENGTH)
):
    """Add a symbol to the watchlist."""
    result = watchlist_service.add_symbol(symbol)

    if not result["success"]:
        raise HTTPException(status_code=400, detail=result["message"])

    return SymbolActionResponse(
        success=True,
        message=result["message"],
        symbols=watchlist_service.get_watchlist()
    )


@router.delete("/{symbol}", response_model=SymbolActionResponse)
async def remove_from_watchlist(
    symbol: str = Path(..., min_length=1, max_length=MAX_SYMBOL_LENGTH)
):
    """Remove a symbol from the watchlist."""
    result = watchlist_service.remove_symbol(symbol)

    if not result["success"]:
        raise HTTPException(status_code=400, detail=result["message"])

    return SymbolActionResponse(
        success=True,
        message=result["message"],
        symbols=watchlist_service.get_watchlist()
    )


@router.get("/search", response_model=list[SearchResult])
async def search_symbols(
    q: str = Query(..., min_length=1, description="Search query for stock symbol")
):
    """Search for stock symbols."""
    results = watchlist_service.search_symbols(q)
    return [SearchResult(**r) for r in results]


@router.get("/validate/{symbol}")
async def validate_symbol(
    symbol: str = Path(..., min_length=1, max_length=MAX_SYMBOL_LENGTH)
):
    """Check if a symbol is valid."""
    is_valid = watchlist_service.validate_symbol(symbol)
    return {
        "symbol": symbol.upper(),
        "valid": is_valid,
        "in_watchlist": watchlist_service.is_in_watchlist(symbol)
    }
